acquire_range.py

```diff
@@ -47,28 +72,46 @@ def __init__(self):
     @prop
     def _func_bar():
         pass
+
+class foo9:
+    try:
+        foo()
+    except:
+        bar()
+
+class foo10:
+    for x in [1]:
+        print(x)
+    else:
+        print("Foo")
+
+class foo11:
+    x = [1]
+    while x:
+        x.pop
+    else:
+        print("Foo")
 '''
 
 # TODO: Add tests and probably code for
 # AsyncFunctionDef, AsyncWith, AsyncFor
+ENTRY_POINTS = ast.ClassDef, ast.FunctionDef
 
 tree = ast.parse(source)
 for entry in tree.body:
-    if isinstance(entry, (ast.ClassDef, ast.FunctionDef)):
+    if isinstance(entry, ENTRY_POINTS):
         print('Body', entry.body)
         last_body = entry.body[-1]
-        while isinstance(last_body, (ast.ClassDef, ast.For, ast.FunctionDef, ast.If, ast.While, ast.With)):
-            print('Available fields:', dir(last_body))
+        while isinstance(last_body, tuple(TRAVERSABLE_FIELDS.keys())):
+            fields = TRAVERSABLE_FIELDS.get(last_body.__class__, ())
+            print('Available fields:', fields)
             print('Body', last_body.body)
-            # TODO: For some ast node types this is too simple. For
-            # example for try-except, we will have to
-            # check for ast.Try.finalbody, ast.Try.orelse and
-            # ast.Try.handlers sequentially and move to the first defined
-            # attribute among these.
-            # Similarly for if-else. For Python 2 there is no ast.Try,
-            # but only ast.TryExcept and ast.TryFinally. The code here
-            # only works for ast nodes that only have one child "body".
-            last_body = last_body.body[-1]
+            for field in reversed(fields):
+                child = getattr(last_body, field)
+                if child != []:
+                    break
+            print("Child:", child)
+            last_body = child[-1]
```

Comment:
`child` is actually the list containing references to all the children node.
I think the confusion is because of my nomenclature 😛 , I think I should refactor it like this:
```python
for field in reversed(fields):
    children = getattr(last_body, field)
    if children:
        break
    last_body = children[-1] # The last child
```